src/main/scala/org/wikimedia/analytics/hdfstools/HdfsRsyncExec.scala

```diff
@@ -257,91 +237,83 @@ class HdfsRsyncExec(config: HdfsRsyncConfig) {
+                    copyNewFile(src, target.get)
                 } else if (config.ignoreTimes || // Force copy/update
                     src.getLen != foundTarget.get.getLen || // Size is different
                     (!config.sizeOnly &&                    // Use modificationTime
                         approxDiffTimestamps(src, foundTarget.get))) {
-                    copyUpdate(src, target, copy)
+                    updateFile(src, target.get)
                 } else {
-                    skip(src, target, copy)
+                    skipFile(src, target.get)
                 }
             }
-            preserveModificationTimesAsNeeded(src, targetToUpdate)
-            updatePermsAsNeeded(src, targetToUpdate, isDir = src.isDirectory, isNew)
-            Some(target)
+            if (config.preserveTimes) {
+                preserveModificationTimes(src, targetToUpdate)
+            }
+            // Update perms if needed
+            if (config.preservePerms || (isNew && (isNew && config.chmodFiles.isDefined))) {
+                updateFilePerms(src, targetToUpdate, config.chmodFiles)
+            }
         }
     }
 
     /**
-     * Function listing files and folders of a src folder or glob (first call before recursion),
-     * and building a Map[filename, FileStatus] from it.
+     * Function processing a single folder with a target and foundTarget.
+     * The target folder is created if it doesn't exist, and its permissions are
+     * updated depending on configuration.
      *
-     * @param srcPath the src path to scan
-     * @param isBasePath whether to use a glob if true or a folder scan if false
-     * @return the Map[filename, FileStatus] of the folder/glob content
+     * @param target the target where to create the new folderif necessary (if any)
+     * @param foundTarget the found target in dst (same filename as src) if any
+     * @return Some path if dst was defined (can be a non-existent file in case of dryrun),
+     *         None if no dst was provided (logs only).
      */
-    private def getSrcFilesAndFolders(srcPath: Path, isBasePath: Boolean): Map[String, FileStatus] = {
-        // Use globStatus only for recursive tree root, as later we'll always have folders.
-        // This prevent having to use the /* addition to Path for glob to return content folders.
-        if (isBasePath) {
-            val globSrc = config.srcFs.globStatus(srcPath)
-            val globSrcChecked = if (globSrc != null) globSrc.toSeq else Seq.empty
-            globSrcChecked.map(f => (f.getPath.getName, f)).toMap
-        } else {
-            config.srcFs.listStatus(srcPath).map(f => (f.getPath.getName, f)).toMap
+    private def processDir(
+        target: Option[Path],
+        foundTarget: Option[FileStatus]
+    ): Unit = {
+        // Don't log dir info if no dest provided, only file
+        if (target.isDefined) {
+            val isNew = foundTarget.isEmpty
+            val targetToUpdate = {
```

Comment:
Since createDir returns Either[Path, FileStatus], targetToUpdate might be a Left[Path]?, Oh, I guess that is ok since updateDirPerms also takes and Either and will assume dry run if Left[Path]?